hoomd/simulation.py

```diff
@@ -129,15 +129,21 @@ def create_state_from_snapshot(self, snapshot):
         When `timestep` is `None` before calling, `create_state_from_snapshot`
         sets `timestep` to 0.
 
-        Warning:
-            *snapshot* must be a `hoomd.Snapshot`. Use `create_state_from_gsd`
-            to read GSD files. `create_state_from_snapshot` does not support
-            ``gsd.hoomd.Snapshot`` objects from the ``gsd`` Python package.
+        Note:
+            *snapshot* must be either a `hoomd.Snapshot` or a
+            ``gsd.hoomd.Snapshot``. A ``gsd.hoomd.Snapshot`` will first be
+            converted to a `hoomd.Snapshot`.
         """
         if self.state is not None:
             raise RuntimeError("Cannot initialize more than once\n")
 
-        self._state = State(self, snapshot)
+        try:
+            # snapshot is hoomd.Snapshot
+            self._state = State(self, snapshot)
+        except AttributeError:
```

Comment:
There a lot of ways for `AttributeError`s to be raised that do not imply the type is a `gsd.hoomd.Snapshot`. I'd prefer to see more explicit type checks and to raise a `TypeError` when the caller provides an invalid type.

Reply:
The way I'm thinking to do type checking is like this:
```
import gsd.hoomd
isinstance(snapshot, gsd.hoomd.Snapshot)
```
Is there a better way to do this that won't add gsd as a dependency?

Reply:
Better is subjective here. What we can do is look at the name of the object something like

```python
def _match_class_path(obj, *matches):
     return any(cls.__module__ + '.' + cls.__name__ in matches
                       for cls in inspect.getmro(type(obj)))
# Example usage
if _match_class_path(system, 'gsd.hoomd.Snapshot'):
    pass        
```
would work without introducing even soft dependencies (this is what freud does to support so many ways to define a system.

Reply:
Oh that's very cool! Thank you!